jung-algorithms/src/main/java/edu/uci/ics/jung/layout/algorithms/FRBHVisitorLayoutAlgorithm.java

```diff
@@ -132,7 +138,13 @@ public synchronized void step() {
     }
     Graph<N> graph = layoutModel.getGraph();
     currentIteration++;
+
+    // the cost of building the tree each time is less than the O(n^2) cost in the repulsion part
```

Comment:
can the tree be incrementally updated?

Reply:
Good question. I do incremental updates of the spatial strucures on the viz side, but that is because they have more to do with 'can i see it or not' instead of a distance measurement to a point. You would certainly lose some accuracy if you start estimating there too. I already suspect that the substitution of the force vector for the actual position results in some clumping of the nodes. You can set EPSILON to zero and force it to visit all nodes. I think that should give the same result as the original algorithm. Hmm, I just thought of a new unit test to add....

Reply:
I saw some discussion here: https://www.reddit.com/r/programming/comments/i75hu/gorgeous_description_of_the_barneshut_algorithm/
that implies that updating or rebuilding the tree are both NlogN

Reply:
related: the repulsion part isn't actually still O(n^2), is it?  the whole point of the spatial data structure is so that you don't have to do all pairwise comparisons directly, I thought.

Reply:
with the BHQT, the replusion part should be O(nlogn) 
without the BHQT (existing code) if is O(n^2)

Reply:
Ah, I see.  It's not clear from the comment that that's what you're getting at.